예시를 들어 행하는 것이 좋을 듯 하여 자주 쓰는 것으로 보이는 refFlat.txt를 대상으로 해보겠습니다

In [1]:
ref_path = './refFlat.txt'

먼저, 해당 파일을 불러오고 싶은데 이는 몇 가지 방법이 있습니다.  
첫번째 방법은 그냥 파일 열어서 가져오는 것입니다

In [2]:
with open(ref_path) as refFlat:
    for i in range(5):
        rl = refFlat.readline()
        print(rl)
        print(rl.split('\t'))
        print('-'*100)

DDX11L1	NR_046018	chr1	+	11873	14409	14409	14409	3	11873,12612,13220,	12227,12721,14409,

['DDX11L1', 'NR_046018', 'chr1', '+', '11873', '14409', '14409', '14409', '3', '11873,12612,13220,', '12227,12721,14409,\n']
----------------------------------------------------------------------------------------------------
WASH7P	NR_024540	chr1	-	14361	29370	29370	29370	11	14361,14969,15795,16606,16857,17232,17605,17914,18267,24737,29320,	14829,15038,15947,16765,17055,17368,17742,18061,18366,24891,29370,

['WASH7P', 'NR_024540', 'chr1', '-', '14361', '29370', '29370', '29370', '11', '14361,14969,15795,16606,16857,17232,17605,17914,18267,24737,29320,', '14829,15038,15947,16765,17055,17368,17742,18061,18366,24891,29370,\n']
----------------------------------------------------------------------------------------------------
MIR6859-1	NR_106918	chr1	-	17368	17436	17436	17436	1	17368,	17436,

['MIR6859-1', 'NR_106918', 'chr1', '-', '17368', '17436', '17436', '17436', '1', '17368,', '17436,\n']
-----

만일 gz파일로 압축되어있는걸 그대로 부르고 싶다면 gzip 모듈을 이용함과 동시에 유니코드 변환을 해줘야합니다

In [3]:
import gzip
ref_path_gz = './refFlat.txt.gz'
with gzip.open(ref_path_gz) as refFlat:
    for i in range(5):
        rl = refFlat.readline().decode('UTF-8')
        print(rl)
        print(rl.split('\t'))
        print('-'*100)

DDX11L1	NR_046018	chr1	+	11873	14409	14409	14409	3	11873,12612,13220,	12227,12721,14409,

['DDX11L1', 'NR_046018', 'chr1', '+', '11873', '14409', '14409', '14409', '3', '11873,12612,13220,', '12227,12721,14409,\n']
----------------------------------------------------------------------------------------------------
WASH7P	NR_024540	chr1	-	14361	29370	29370	29370	11	14361,14969,15795,16606,16857,17232,17605,17914,18267,24737,29320,	14829,15038,15947,16765,17055,17368,17742,18061,18366,24891,29370,

['WASH7P', 'NR_024540', 'chr1', '-', '14361', '29370', '29370', '29370', '11', '14361,14969,15795,16606,16857,17232,17605,17914,18267,24737,29320,', '14829,15038,15947,16765,17055,17368,17742,18061,18366,24891,29370,\n']
----------------------------------------------------------------------------------------------------
MIR6859-1	NR_106918	chr1	-	17368	17436	17436	17436	1	17368,	17436,

['MIR6859-1', 'NR_106918', 'chr1', '-', '17368', '17436', '17436', '17436', '1', '17368,', '17436,\n']
-----

보는 바와 같이 줄대로 읽어서 \t로 나누고 여러 작업을 동시에 행할 수 있습니다.  
다만 작업을 행하지 않고 단순불러오기만 할거라면 좋은 방식은 아닙니다

In [4]:
import numpy as np
refFlat_np = np.loadtxt(ref_path,delimiter='\t',dtype=str)
refFlat_np

array([['DDX11L1', 'NR_046018', 'chr1', ..., '3', '11873,12612,13220,',
        '12227,12721,14409,'],
       ['WASH7P', 'NR_024540', 'chr1', ..., '11',
        '14361,14969,15795,16606,16857,17232,17605,17914,18267,24737,29320,',
        '14829,15038,15947,16765,17055,17368,17742,18061,18366,24891,29370,'],
       ['MIR6859-1', 'NR_106918', 'chr1', ..., '1', '17368,', '17436,'],
       ...,
       ['GRB10', 'NM_001350816', 'chr7', ..., '18',
        '50590068,50595436,50603997,50604310,50605289,50606336,50612740,50614769,50616209,50618070,50619169,50626821,50669721,50674435,50703820,50710874,50755886,50792441,',
        '50593098,50595530,50604085,50604377,50605406,50606414,50612839,50614880,50616347,50618139,50619285,50626978,50669863,50674658,50703908,50710990,50756056,50792874,'],
       ['RAB11FIP3', 'NM_001142272', 'chr16', ..., '12',
        '474923,482524,488850,496823,503003,505523,510659,518942,519753,520121,520458,520725,',
        '475045,482736,489000,496859,503097,505627,

numpy loadtxt는 줄별로 읽으며 필요한 수치로 나누어 불러오는 것이 가능합니다. 다만 이는 array 형식이며 테이블 데이터에 최적화되어있지는 않습니다.  
numpy는 애초에 선형수학의 도구로 쓰이는 편이니, 그에 맞게 조정하는데 시간이 쓸릴 가능성이 큽니다

In [5]:
import pandas as pd
col_name = ['transName','transID','chrom','strand','txnSta','txnEnd','cdsSta','cdsEnd','exnNum','exnSta','exnEnd']
refFlat_DF = pd.read_csv('./refFlat.txt', sep='\t', names=col_name, header=None)
refFlat_DF

transName       transID                 chrom strand    txnSta  \
0        DDX11L1     NR_046018                  chr1      +     11873   
1         WASH7P     NR_024540                  chr1      -     14361   
2      MIR6859-1     NR_106918                  chr1      -     17368   
3      MIR6859-2     NR_107062                  chr1      -     17368   
4      MIR6859-3     NR_107063                  chr1      -     17368   
...          ...           ...                   ...    ...       ...   
82077      GRB10  NM_001001555                  chr7      -  50590067   
82078      GRB10  NM_001371008                  chr7      -  50590067   
82079      GRB10  NM_001350816                  chr7      -  50590068   
82080  RAB11FIP3  NM_001142272                 chr16      +    474923   
82081  RAB11FIP3  NM_001142272  chr16_KI270854v1_alt      +     75229   

         txnEnd    cdsSta    cdsEnd  exnNum  \
0         14409     14409     14409       3   
1         29370     29370     29370      11   
2         17436     17436     17436       1   
3         17436     17436     17436       1   
4         17436     17436     17436       1   
...         ...       ...       ...     ...   
82077  50793453  50592951  50674623      17   
82078  50793453  50592951  50674623      16   
82079  50792874  50592951  50674623      18   
82080    523011    482551    520839      12   
82081    125679     82856    123507      12   

                                                  exnSta  \
0                                     11873,12612,13220,   
1      14361,14969,15795,16606,16857,17232,17605,1791...   
2                                                 17368,   
3                                                 17368,   
4                                                 17368,   
...                                                  ...   
82077  50590067,50595436,50603997,50604310,50605289,5...   
82078  50590067,50595436,50603997,50604310,50605289,5...   
82079  50590068,50595436,50603997,50604310,50605289,5...   
82080  474923,482524,488850,496823,503003,505523,5106...   
82081  75229,82829,89154,99694,105890,108215,113279,1...   

                                                  exnEnd  
0                                     12227,12721,14409,  
1      14829,15038,15947,16765,17055,17368,17742,1806...  
2                                                 17436,  
3                                                 17436,  
4                                                 17436,  
...                                                  ...  
82077  50593098,50595530,50604085,50604377,50605406,5...  
82078  50593098,50595530,50604085,50604377,50605406,5...  
82079  50593098,50595530,50604085,50604377,50605406,5...  
82080  475045,482736,489000,496859,503097,505627,5108...  
82081  75351,83041,89304,99730,105984,108319,113420,1...  

[82082 rows x 11 columns]

이 경우 다른건 다 좋으나 일부 문자열인 부분이 마음에 걸립니다

In [6]:
refFlat_DF[['exnSta', 'exnEnd']]

exnSta  \
0                                     11873,12612,13220,   
1      14361,14969,15795,16606,16857,17232,17605,1791...   
2                                                 17368,   
3                                                 17368,   
4                                                 17368,   
...                                                  ...   
82077  50590067,50595436,50603997,50604310,50605289,5...   
82078  50590067,50595436,50603997,50604310,50605289,5...   
82079  50590068,50595436,50603997,50604310,50605289,5...   
82080  474923,482524,488850,496823,503003,505523,5106...   
82081  75229,82829,89154,99694,105890,108215,113279,1...   

                                                  exnEnd  
0                                     12227,12721,14409,  
1      14829,15038,15947,16765,17055,17368,17742,1806...  
2                                                 17436,  
3                                                 17436,  
4                                                 17436,  
...                                                  ...  
82077  50593098,50595530,50604085,50604377,50605406,5...  
82078  50593098,50595530,50604085,50604377,50605406,5...  
82079  50593098,50595530,50604085,50604377,50605406,5...  
82080  475045,482736,489000,496859,503097,505627,5108...  
82081  75351,83041,89304,99730,105984,108319,113420,1...  

[82082 rows x 2 columns]

쉼표를 기준으로 숫자를 구분해야 하므로 맵핑 함수를 씁니다

In [7]:
refFlat_DF['exnSta'] = refFlat_DF['exnSta'].apply(np.fromstring, sep=',', dtype=int)
refFlat_DF['exnEnd'] = refFlat_DF['exnEnd'].apply(np.fromstring, sep=',', dtype=int)
refFlat_DF[['exnSta','exnEnd']]

exnSta  \
0                                  [11873, 12612, 13220]   
1      [14361, 14969, 15795, 16606, 16857, 17232, 176...   
2                                                [17368]   
3                                                [17368]   
4                                                [17368]   
...                                                  ...   
82077  [50590067, 50595436, 50603997, 50604310, 50605...   
82078  [50590067, 50595436, 50603997, 50604310, 50605...   
82079  [50590068, 50595436, 50603997, 50604310, 50605...   
82080  [474923, 482524, 488850, 496823, 503003, 50552...   
82081  [75229, 82829, 89154, 99694, 105890, 108215, 1...   

                                                  exnEnd  
0                                  [12227, 12721, 14409]  
1      [14829, 15038, 15947, 16765, 17055, 17368, 177...  
2                                                [17436]  
3                                                [17436]  
4                                                [17436]  
...                                                  ...  
82077  [50593098, 50595530, 50604085, 50604377, 50605...  
82078  [50593098, 50595530, 50604085, 50604377, 50605...  
82079  [50593098, 50595530, 50604085, 50604377, 50605...  
82080  [475045, 482736, 489000, 496859, 503097, 50562...  
82081  [75351, 83041, 89304, 99730, 105984, 108319, 1...  

[82082 rows x 2 columns]

만일 갯수를 세고 싶다면 apply를 다시 이용하면 됩니다

In [8]:
refFlat_DF['exnSta'].apply(len)

0         3
1        11
2         1
3         1
4         1
         ..
82077    17
82078    16
82079    18
82080    12
82081    12
Name: exnSta, Length: 82082, dtype: int64

exnSta와 exnEnd를 하나씩 짝짓고 싶으면 다음과 같은 방식으로 하면 됩니다  
exnSta와 exnEnd에서 순차적으로 가져온 리스트가 각각 x,y로 지정되고  
x,y를 zip[리스트 내 쌍을 튜플로 묶어 하나씩 반환]으로 한 뒤 리스트로 맞추고 이를 적용한다는 map을 이용  
마지막으로 row 당 하나의 리스트로 만들려고 마지막에 list를 적용하면 됩니다

In [9]:
print(refFlat_DF['exnSta'].loc[0], refFlat_DF['exnEnd'].loc[0])
list(zip(refFlat_DF['exnSta'].loc[0], refFlat_DF['exnEnd'].loc[0]))

[11873 12612 13220] [12227 12721 14409]


[(11873, 12227), (12612, 12721), (13220, 14409)]

In [10]:
list(map(lambda x,y : list(zip(x,y)), refFlat_DF['exnSta'], refFlat_DF['exnEnd']))

[[(11873, 12227), (12612, 12721), (13220, 14409)],
 [(14361, 14829),
  (14969, 15038),
  (15795, 15947),
  (16606, 16765),
  (16857, 17055),
  (17232, 17368),
  (17605, 17742),
  (17914, 18061),
  (18267, 18366),
  (24737, 24891),
  (29320, 29370)],
 [(17368, 17436)],
 [(17368, 17436)],
 [(17368, 17436)],
 [(17368, 17436)],
 [(30365, 30503)],
 [(30365, 30503)],
 [(30365, 30503)],
 [(30365, 30503)],
 [(34610, 35174), (35276, 35481), (35720, 36081)],
 [(34610, 35174), (35276, 35481), (35720, 36081)],
 [(34610, 35174), (35276, 35481), (35720, 36081)],
 [(12892895, 12893474),
  (12894585, 12895164),
  (12895560, 12895872),
  (12898215, 12898270)],
 [(134772, 139696), (139789, 139847), (140074, 140566)],
 [(182387, 182746), (183131, 183240), (183739, 184878)],
 [(187890, 187958)],
 [(187890, 187958)],
 [(187890, 187958)],
 [(187890, 187958)],
 [(205128, 205690), (205792, 205995), (206236, 206597)],
 [(490755, 494898), (494991, 495049), (495276, 495445)],
 [(490755, 494898), (494991, 495049)

In [11]:
refFlat_DF['exnList'] = list(map(lambda x,y : list(zip(x,y)), refFlat_DF['exnSta'], refFlat_DF['exnEnd']))
refFlat_DF['exnList']

0         [(11873, 12227), (12612, 12721), (13220, 14409)]
1        [(14361, 14829), (14969, 15038), (15795, 15947...
2                                         [(17368, 17436)]
3                                         [(17368, 17436)]
4                                         [(17368, 17436)]
                               ...                        
82077    [(50590067, 50593098), (50595436, 50595530), (...
82078    [(50590067, 50593098), (50595436, 50595530), (...
82079    [(50590068, 50593098), (50595436, 50595530), (...
82080    [(474923, 475045), (482524, 482736), (488850, ...
82081    [(75229, 75351), (82829, 83041), (89154, 89304...
Name: exnList, Length: 82082, dtype: object

numpy계산이 더 편하다 하면 이와 같은 방법도 있습니다

In [12]:
refFlat_DF['exnList'] = list(map(lambda x,y : np.transpose([x,y]), refFlat_DF['exnSta'], refFlat_DF['exnEnd']))
refFlat_DF['exnList']

0         [[11873, 12227], [12612, 12721], [13220, 14409]]
1        [[14361, 14829], [14969, 15038], [15795, 15947...
2                                         [[17368, 17436]]
3                                         [[17368, 17436]]
4                                         [[17368, 17436]]
                               ...                        
82077    [[50590067, 50593098], [50595436, 50595530], [...
82078    [[50590067, 50593098], [50595436, 50595530], [...
82079    [[50590068, 50593098], [50595436, 50595530], [...
82080    [[474923, 475045], [482524, 482736], [488850, ...
82081    [[75229, 75351], [82829, 83041], [89154, 89304...
Name: exnList, Length: 82082, dtype: object

exnList에서 각각의 길이를 알고 싶으면 차이값을 구하면 되므로 np.diff함수를 쓰면 됩니다  
또한 길이의 합은 sum함수로 구하면 편합니다

In [13]:
refFlat_DF['exnLenList'] = refFlat_DF['exnList'].apply(np.diff)
refFlat_DF['exnLen'] = refFlat_DF['exnLenList'].apply(sum)
refFlat_DF[['exnLenList','exnLen']]

exnLenList  exnLen
0                                 [[354], [109], [1189]]  [1652]
1      [[468], [69], [152], [159], [198], [136], [137...  [1769]
2                                                 [[68]]    [68]
3                                                 [[68]]    [68]
4                                                 [[68]]    [68]
...                                                  ...     ...
82077  [[3031], [94], [88], [67], [117], [78], [99], ...  [5018]
82078  [[3031], [94], [88], [67], [117], [78], [99], ...  [4857]
82079  [[3030], [94], [88], [67], [117], [78], [99], ...  [5336]
82080  [[122], [212], [150], [36], [94], [104], [141]...  [3662]
82081  [[122], [212], [150], [36], [94], [104], [141]...  [3662]

[82082 rows x 2 columns]

apply에 커스텀 프로세스로 복잡한 계산을 하고 싶다면 우선 함수를 정의해야합니다  
이에 예시로는 cdsSta 및 cdsEnd를 기준으로 exnList를 나누는 것을 해보겠습니다  
편의 상 numpy array로 합니다

In [14]:
# 입력 값은 refFlat_DF의 strand, exnSta, exnEnd, exnList, cdsSta, cdsEnd
def sep_exn(F_strand, F_exnSta, F_exnEnd, F_exnList, F_cdsSta, F_cdsEnd):
    # intron 구하기
    F_intron = np.transpose([F_exnEnd[:-1], F_exnSta[1:]])
    # cdsSta와 cdsEnd가 같으면 패스
    if F_cdsSta != F_cdsEnd:
        # exn에서 cds의 위치 파악
        front_index = np.searchsorted(F_exnEnd,F_cdsSta)+1
        back__index = np.searchsorted(F_exnSta,F_cdsEnd)+1
        # cds를 두개 씩 아래에서 쑤셔넣고 정렬하여 나누기 쉽게
        F_exnList = np.sort(np.concatenate((F_exnList, [[F_cdsSta,F_cdsSta],[F_cdsEnd,F_cdsEnd]])),axis=0)
        # cds 중간 추출
        F_middl_lst = F_exnList[front_index:back__index]
        # strand에 따라 앞뒤 추출
        if F_strand == '+':
            F_front_lst = F_exnList[:front_index]
            F_back__lst = F_exnList[back__index:]
        else:
            F_back__lst = F_exnList[:front_index]
            F_front_lst = F_exnList[back__index:]
        # 나눈 리스트에서 잘라지지 않을 경우에 대해 보정 및 각각의 길이랑 전체길이
        def LLS_cal(FF_lst):
            FF_len = np.diff(FF_lst)
            FF_TF = FF_len.astype(bool).flatten()
            FF_lst = FF_lst[FF_TF]
            FF_len = FF_len[FF_TF]
            FF_sum = sum(FF_len)
            return FF_len, FF_lst, FF_sum
        #함수 적용
        F_front_len, F_front_lst, F_front_sum = LLS_cal(F_front_lst)
        F_middl_len, F_middl_lst, F_middl_sum = LLS_cal(F_middl_lst)
        F_back__len, F_back__lst, F_back__sum = LLS_cal(F_back__lst)
    # 같을 경우는 중간만 입력받고 내뱉기
    else:
        F_front_lst, F_back__lst = [],[]
        F_middl_lst = F_exnList
        F_middl_len = np.diff(F_middl_lst)
        F_middl_sum = sum(F_middl_len)
        F_front_len, F_back__len = [],[]
        F_front_sum, F_back__sum = 0, 0
    # 처리 결과를 리스트로 반환
    return [F_front_lst, F_front_len, F_front_sum, F_middl_lst, F_middl_len, F_middl_sum, F_back__lst, F_back__len, F_back__sum, F_intron]

예시를 위해 하나를 찝었습니다

In [15]:
ex_ref_DF = refFlat_DF.loc[60]
print(ex_ref_DF)
F_strand, F_exnSta, F_exnEnd, F_exnList, F_cdsSta, F_cdsEnd = ex_ref_DF['strand'], ex_ref_DF['exnSta'], ex_ref_DF['exnEnd'], ex_ref_DF['exnList'], ex_ref_DF['cdsSta'], ex_ref_DF['cdsEnd']

transName                                                 NLRP2
transID                                               NM_017852
chrom                                      chr19_GL949752v1_alt
strand                                                        +
txnSta                                                   869115
txnEnd                                                   903919
cdsSta                                                   872770
cdsEnd                                                   903680
exnNum                                                       13
exnSta        [869115, 872753, 877254, 880503, 884366, 88491...
exnEnd        [869209, 873050, 877299, 880575, 884432, 88647...
exnList       [[869115, 869209], [872753, 873050], [877254, ...
exnLenList    [[94], [297], [45], [72], [66], [1567], [171],...
exnLen                                                   [3539]
Name: 60, dtype: object


먼저 intron을 구하려면 (1,2) (3,4) 등에서 (2,3)으로 묶으면 되므로  
간단히 생각하면 1,4 등의 맨처음과 마지막 빼고 (3,X), (X,2)로 다시 엮으면 되는 것입니다

In [16]:
print(F_exnSta)
print(F_exnEnd)

[869115 872753 877254 880503 884366 884911 887830 888934 892808 893288
 897050 900098 903541]
[869209 873050 877299 880575 884432 886478 888001 889099 892979 893459
 897221 900269 903919]


즉, exnSta의 맨 앞 빼고, exnEnd의 맨 뒤 빼고 이를 엮으면 됩니다

In [17]:
print(ex_ref_DF['exnSta'][1:])
print(ex_ref_DF['exnEnd'][:-1])

F_intron = np.transpose([ex_ref_DF['exnEnd'][:-1], ex_ref_DF['exnSta'][1:]])
print(F_intron)

[872753 877254 880503 884366 884911 887830 888934 892808 893288 897050
 900098 903541]
[869209 873050 877299 880575 884432 886478 888001 889099 892979 893459
 897221 900269]
[[869209 872753]
 [873050 877254]
 [877299 880503]
 [880575 884366]
 [884432 884911]
 [886478 887830]
 [888001 888934]
 [889099 892808]
 [892979 893288]
 [893459 897050]
 [897221 900098]
 [900269 903541]]


exnList에서 cds가 어느 위치에 있는지를 확인해보려면 np.searchsorted를 이용하면 편합니다  
다만 1D의 경우만 찾을 수 있고, cds가 exn 내부를 가를 경우를 상정하면 cdsEnd는 exnSta랑, cdsSta는 exnEnd랑 비교하는 것이 좋습니다  
우선 Sta는 Sta대로 End는 End대로 비교해보겠습니다

In [18]:
front_index = np.searchsorted(F_exnSta,F_cdsSta)
back__index = np.searchsorted(F_exnEnd,F_cdsEnd)
print(front_index, back__index)

2 12


이것을 기준으로 나눠보면 다음과 같습니다

In [19]:
print(F_exnList[:front_index])
print('---',F_cdsSta,'---')
print(F_exnList[front_index:back__index])
print('---',F_cdsEnd,'---')
print(F_exnList[back__index:])

[[869115 869209]
 [872753 873050]]
--- 872770 ---
[[877254 877299]
 [880503 880575]
 [884366 884432]
 [884911 886478]
 [887830 888001]
 [888934 889099]
 [892808 892979]
 [893288 893459]
 [897050 897221]
 [900098 900269]]
--- 903680 ---
[[903541 903919]]


 [872753, 873050]] #요걸 872770로 분할  
--- 872770 --- <- 나뉜 부분  
[[877254, 877299]  

이것을

 [872753, 872770]] #872770을 끝으로 재설정  
--- 872770 --- <- 나뉜 부분  
[[872770, 873050] #872770을 시작으로 재설정  
 [877254, 877299]  
 이렇게 872770을 기준으로 잘라 붙이고 싶으므로 이걸 간단히 해결하기 위해서는 그냥 872770을 추가시켜서 다시 배열을 만들면 됩니다

In [20]:
ex_exnList = np.concatenate((F_exnList, [[F_cdsSta, F_cdsSta], [F_cdsEnd, F_cdsEnd]]))
ex_exnList

array([[869115, 869209],
       [872753, 873050],
       [877254, 877299],
       [880503, 880575],
       [884366, 884432],
       [884911, 886478],
       [887830, 888001],
       [888934, 889099],
       [892808, 892979],
       [893288, 893459],
       [897050, 897221],
       [900098, 900269],
       [903541, 903919],
       [872770, 872770],
       [903680, 903680]])

이러면 맨 끝에 [[cdsSta, cdsSta], [cdsEnd, cdsEnd]]의 배열이 붙었고 이걸 위에서 부터 내림차순으로 정렬하면


In [21]:
F_exnList = np.sort(ex_exnList,axis=0)
F_exnList

array([[869115, 869209],
       [872753, 872770],
       [872770, 873050],
       [877254, 877299],
       [880503, 880575],
       [884366, 884432],
       [884911, 886478],
       [887830, 888001],
       [888934, 889099],
       [892808, 892979],
       [893288, 893459],
       [897050, 897221],
       [900098, 900269],
       [903541, 903680],
       [903680, 903919]])

이와 같이 정렬되어 872770와 903680을 기준으로 나누면 됩니다

In [22]:
print(F_exnList[:front_index])
print('---',F_cdsSta,'---')
print(F_exnList[front_index:back__index])
print('---',F_cdsEnd,'---')
print(F_exnList[back__index:])

[[869115 869209]
 [872753 872770]]
--- 872770 ---
[[872770 873050]
 [877254 877299]
 [880503 880575]
 [884366 884432]
 [884911 886478]
 [887830 888001]
 [888934 889099]
 [892808 892979]
 [893288 893459]
 [897050 897221]]
--- 903680 ---
[[900098 900269]
 [903541 903680]
 [903680 903919]]


보는 바와 같이 두 열이 추가되어서 제대로 나눠지지 않으므로 이걸 사전에 해결하기 위해 searchsorted를 다음과 같이 설정합니다

In [23]:
front_index = np.searchsorted(F_exnEnd, F_cdsSta)+1
back__index = np.searchsorted(F_exnSta, F_cdsEnd)+1
print(front_index, back__index)

2 14


In [24]:
print(F_exnList[:front_index])
print('---',F_cdsSta,'---')
print(F_exnList[front_index:back__index])
print('---',F_cdsEnd,'---')
print(F_exnList[back__index:])

[[869115 869209]
 [872753 872770]]
--- 872770 ---
[[872770 873050]
 [877254 877299]
 [880503 880575]
 [884366 884432]
 [884911 886478]
 [887830 888001]
 [888934 889099]
 [892808 892979]
 [893288 893459]
 [897050 897221]
 [900098 900269]
 [903541 903680]]
--- 903680 ---
[[903680 903919]]


이걸 토대로 해서 나눈 것이 다음 방식입니다

In [25]:
F_middl_lst = F_exnList[front_index:back__index]
if F_strand == '+':
    F_front_lst = F_exnList[:front_index]
    F_back__lst = F_exnList[back__index:]
else:
    F_back__lst = F_exnList[:front_index]
    F_front_lst = F_exnList[back__index:]

다만 만약 [(1,2)(4,5)(7,8)]일 경우 분할이 3, 6일 시  
[(1,2)(3,3)(4,5)(6,6)(7,8)]이 될 수 있으므로 이를 해결하기 위해 len으로 길이를 구하고 그것이 0일 경우 boolen으로 제거하는 방식을 포함한 것이

In [26]:
def LLS_cal(FF_lst):
    FF_len = np.diff(FF_lst)              # 길이 구하기
    FF_TF = FF_len.astype(bool).flatten() # 길이가 0이면 False
    FF_lst = FF_lst[FF_TF]                # 길이가 False인 경우 제거
    FF_len = FF_len[FF_TF]                # 길이가 False인 경우 제거
    FF_sum = sum(FF_len)                  # 길이 합
    return FF_len, FF_lst, FF_sum

위 함수입니다

In [27]:
F_front_len, F_front_lst, F_front_sum = LLS_cal(F_front_lst)
F_middl_len, F_middl_lst, F_middl_sum = LLS_cal(F_middl_lst)
F_back__len, F_back__lst, F_back__sum = LLS_cal(F_back__lst)

In [28]:
ex_ref_DF = [F_front_lst, F_front_len, F_front_sum, F_middl_lst, F_middl_len, F_middl_sum, F_back__lst, F_back__len, F_back__sum, F_intron]
pd.DataFrame(ex_ref_DF)

0
0               [[869115, 869209], [872753, 872770]]
1                                       [[94], [17]]
2                                              [111]
3  [[872770, 873050], [877254, 877299], [880503, ...
4  [[280], [45], [72], [66], [1567], [171], [165]...
5                                             [3189]
6                                 [[903680, 903919]]
7                                            [[239]]
8                                              [239]
9  [[869209, 872753], [873050, 877254], [877299, ...

이제 이걸 기존 DataFrame에 적용시켜야합니다  
이를 위해선 lambda를 써야하고 이걸 map으로 적용시켜서 DataFrame로 만들면 됩니다  
변수는 총 6개로 refFlat_DF['strand'], refFlat_DF['exnSta'], refFlat_DF['exnEnd'], refFlat_DF['cdsSta'], refFlat_DF['cdsEnd']  
받을 시 쓸 colums은 ['exnList', 'exnLenList', 'exnLen', 'utr5','utr5LenList','utr5Len','cdsList','cdsLenList','cdsLen','utr3','utr3LenList','utr3Len','intron']로 만듭니다

In [29]:
sep_UTR = pd.DataFrame(map(lambda x1,x2,x3,x4,x5,x6:sep_exn(x1,x2,x3,x4,x5,x6),\
                               refFlat_DF['strand'], refFlat_DF['exnSta'], refFlat_DF['exnEnd'], refFlat_DF['exnList'], refFlat_DF['cdsSta'], refFlat_DF['cdsEnd']),\
                               columns=['utr5','utr5LenList','utr5Len','cdsList','cdsLenList','cdsLen','utr3','utr3LenList','utr3Len','intron'])
sep_UTR

utr5  \
0                                                     []   
1                                                     []   
2                                                     []   
3                                                     []   
4                                                     []   
...                                                  ...   
82077  [[50674623, 50674658], [50703820, 50703908], [...   
82078  [[50674623, 50674658], [50732271, 50732368], [...   
82079  [[50674623, 50674658], [50703820, 50703908], [...   
82080               [[474923, 475045], [482524, 482551]]   
82081                   [[75229, 75351], [82829, 82856]]   

                             utr5LenList utr5Len  \
0                                     []       0   
1                                     []       0   
2                                     []       0   
3                                     []       0   
4                                     []       0   
...                                  ...     ...   
82077         [[35], [88], [170], [230]]   [523]   
82078                [[35], [97], [230]]   [362]   
82079  [[35], [88], [116], [170], [433]]   [842]   
82080                      [[122], [27]]   [149]   
82081                      [[122], [27]]   [149]   

                                                 cdsList  \
0       [[11873, 12227], [12612, 12721], [13220, 14409]]   
1      [[14361, 14829], [14969, 15038], [15795, 15947...   
2                                       [[17368, 17436]]   
3                                       [[17368, 17436]]   
4                                       [[17368, 17436]]   
...                                                  ...   
82077  [[50592951, 50593098], [50595436, 50595530], [...   
82078  [[50592951, 50593098], [50595436, 50595530], [...   
82079  [[50592951, 50593098], [50595436, 50595530], [...   
82080  [[482551, 482736], [488850, 489000], [496823, ...   
82081  [[82856, 83041], [89154, 89304], [99694, 99730...   

                                              cdsLenList  cdsLen  \
0                                 [[354], [109], [1189]]  [1652]   
1      [[468], [69], [152], [159], [198], [136], [137...  [1769]   
2                                                 [[68]]    [68]   
3                                                 [[68]]    [68]   
4                                                 [[68]]    [68]   
...                                                  ...     ...   
82077  [[147], [94], [88], [67], [117], [78], [99], [...  [1611]   
82078  [[147], [94], [88], [67], [117], [78], [99], [...  [1611]   
82079  [[147], [94], [88], [67], [117], [78], [99], [...  [1611]   
82080  [[185], [150], [36], [94], [104], [141], [82],...  [1341]   
82081  [[185], [150], [36], [94], [104], [141], [82],...  [1341]   

                         utr3 utr3LenList utr3Len  \
0                          []          []       0   
1                          []          []       0   
2                          []          []       0   
3                          []          []       0   
4                          []          []       0   
...                       ...         ...     ...   
82077  [[50590067, 50592951]]    [[2884]]  [2884]   
82078  [[50590067, 50592951]]    [[2884]]  [2884]   
82079  [[50590068, 50592951]]    [[2883]]  [2883]   
82080      [[520839, 523011]]    [[2172]]  [2172]   
82081      [[123507, 125679]]    [[2172]]  [2172]   

                                                  intron  
0                       [[12227, 12612], [12721, 13220]]  
1      [[14829, 14969], [15038, 15795], [15947, 16606...  
2                                                     []  
3                                                     []  
4                                                     []  
...                                                  ...  
82077  [[50593098, 50595436], [50595530, 50603997], [...  
82078  [[50593098, 50595436], [50595530, 50603997], [...  


sep_UTR을 refFlat_DF옆에다 붙일 것이므로

In [30]:
refFlat_DF = refFlat_DF.merge(sep_UTR,right_index=True,left_index=True)
refFlat_DF

transName       transID                 chrom strand    txnSta  \
0        DDX11L1     NR_046018                  chr1      +     11873   
1         WASH7P     NR_024540                  chr1      -     14361   
2      MIR6859-1     NR_106918                  chr1      -     17368   
3      MIR6859-2     NR_107062                  chr1      -     17368   
4      MIR6859-3     NR_107063                  chr1      -     17368   
...          ...           ...                   ...    ...       ...   
82077      GRB10  NM_001001555                  chr7      -  50590067   
82078      GRB10  NM_001371008                  chr7      -  50590067   
82079      GRB10  NM_001350816                  chr7      -  50590068   
82080  RAB11FIP3  NM_001142272                 chr16      +    474923   
82081  RAB11FIP3  NM_001142272  chr16_KI270854v1_alt      +     75229   

         txnEnd    cdsSta    cdsEnd  exnNum  \
0         14409     14409     14409       3   
1         29370     29370     29370      11   
2         17436     17436     17436       1   
3         17436     17436     17436       1   
4         17436     17436     17436       1   
...         ...       ...       ...     ...   
82077  50793453  50592951  50674623      17   
82078  50793453  50592951  50674623      16   
82079  50792874  50592951  50674623      18   
82080    523011    482551    520839      12   
82081    125679     82856    123507      12   

                                                  exnSta  ...  \
0                                  [11873, 12612, 13220]  ...   
1      [14361, 14969, 15795, 16606, 16857, 17232, 176...  ...   
2                                                [17368]  ...   
3                                                [17368]  ...   
4                                                [17368]  ...   
...                                                  ...  ...   
82077  [50590067, 50595436, 50603997, 50604310, 50605...  ...   
82078  [50590067, 50595436, 50603997, 50604310, 50605...  ...   
82079  [50590068, 50595436, 50603997, 50604310, 50605...  ...   
82080  [474923, 482524, 488850, 496823, 503003, 50552...  ...   
82081  [75229, 82829, 89154, 99694, 105890, 108215, 1...  ...   

                                                    utr5  \
0                                                     []   
1                                                     []   
2                                                     []   
3                                                     []   
4                                                     []   
...                                                  ...   
82077  [[50674623, 50674658], [50703820, 50703908], [...   
82078  [[50674623, 50674658], [50732271, 50732368], [...   
82079  [[50674623, 50674658], [50703820, 50703908], [...   
82080               [[474923, 475045], [482524, 482551]]   
82081                   [[75229, 75351], [82829, 82856]]   

                             utr5LenList utr5Len  \
0                                     []       0   
1                                     []       0   
2                                     []       0   
3                                     []       0   
4                                     []       0   
...                                  ...     ...   
82077         [[35], [88], [170], [230]]   [523]   
82078                [[35], [97], [230]]   [362]   
82079  [[35], [88], [116], [170], [433]]   [842]   
82080                      [[122], [27]]   [149]   
82081                      [[122], [27]]   [149]   

                                                 cdsList  \
0       [[11873, 12227], [12612, 12721], [13220, 14409]]   
1      [[14361, 14829], [14969, 15038], [15795, 15947...   
2                                       [[17368, 17436]]   
3                                       [[17368, 17436]]   
4                                       [[17368, 17436]]   
...                                                  ...   
82077  [[50592951,

이런식으로 데이터프레임을 조작할 수 있습니다

여기서 transName으로 묶어서 보고싶을 경우는 다음과 같은 방식을 이용하면 편합니다

In [31]:
refFlat_DF[refFlat_DF["transName"] == "GRB10"]

transName       transID chrom strand    txnSta    txnEnd    cdsSta  \
47331     GRB10  NM_001350814  chr7      -  50590067  50782896  50592951   
58153     GRB10  NM_001350815  chr7      -  50590067  50792874  50592951   
82074     GRB10  NM_001001550  chr7      -  50590067  50705301  50592951   
82075     GRB10  NM_001001549  chr7      -  50590067  50732353  50592951   
82076     GRB10  NM_001371009  chr7      -  50590067  50782896  50592951   
82077     GRB10  NM_001001555  chr7      -  50590067  50793453  50592951   
82078     GRB10  NM_001371008  chr7      -  50590067  50793453  50592951   
82079     GRB10  NM_001350816  chr7      -  50590068  50792874  50592951   

         cdsEnd  exnNum                                             exnSta  \
47331  50732322      19  [50590067, 50595436, 50603997, 50604310, 50605...   
58153  50792524      16  [50590067, 50595436, 50603997, 50604310, 50605...   
82074  50674623      16  [50590067, 50595436, 50603997, 50604310, 50605...   
82075  50732322      15  [50590067, 50595436, 50603997, 50604310, 50605...   
82076  50782539      16  [50590067, 50595436, 50603997, 50604310, 50605...   
82077  50674623      17  [50590067, 50595436, 50603997, 50604310, 50605...   
82078  50674623      16  [50590067, 50595436, 50603997, 50604310, 50605...   
82079  50674623      18  [50590068, 50595436, 50603997, 50604310, 50605...   

       ...                                               utr5  \
47331  ...  [[50732322, 50732368], [50755886, 50756056], [...   
58153  ...                             [[50792524, 50792874]]   
82074  ...  [[50674623, 50674658], [50703820, 50703908], [...   
82075  ...                             [[50732322, 50732353]]   
82076  ...                             [[50782539, 50782896]]   
82077  ...  [[50674623, 50674658], [50703820, 50703908], [...   
82078  ...  [[50674623, 50674658], [50732271, 50732368], [...   
82079  ...  [[50674623, 50674658], [50703820, 50703908], [...   

                             utr5LenList utr5Len  \
47331        [[46], [170], [110], [473]]   [799]   
58153                            [[350]]   [350]   
82074                [[35], [88], [162]]   [285]   
82075                             [[31]]    [31]   
82076                            [[357]]   [357]   
82077         [[35], [88], [170], [230]]   [523]   
82078                [[35], [97], [230]]   [362]   
82079  [[35], [88], [116], [170], [433]]   [842]   

                                                 cdsList  \
47331  [[50592951, 50593098], [50595436, 50595530], [...   
58153  [[50592951, 50593098], [50595436, 50595530], [...   
82074  [[50592951, 50593098], [50595436, 50595530], [...   
82075  [[50592951, 50593098], [50595436, 50595530], [...   
82076  [[50592951, 50593098], [50595436, 50595530], [...   
82077  [[50592951, 50593098], [50595436, 50595530], [...   
82078  [[50592951, 50593098], [50595436, 50595530], [...   
82079  [[50592951, 50593098], [50595436, 50595530], [...   

                                              cdsLenList  cdsLen  \
47331  [[147], [94], [88], [67], [117], [78], [99], [...  [1785]   
58153  [[147], [94], [88], [67], [117], [78], [99], [...  [1899]   
82074  [[147], [94], [88], [67], [117], [78], [99], [...  [1611]   
82075  [[147], [94], [88], [67], [117], [78], [99], [...  [1647]   
82076  [[147], [94], [88], [67], [117], [78], [99], [...  [1932]   
82077  [[147], [94], [88], [67], [117], [78], [99], [...  [1611]   
82078  [[147], [94], [88], [67], [117], [78], [99], [...  [1611]   
82079  [[147], [94], [88], [67], [117], [78], [99], [...  [1611]   

                         utr3 utr3LenList utr3Len  \
47331  [[50590067, 50592951]]    [[2884]]  [2884]   
58153  [[50590067, 50592951]]    [[2884]]  [2884]   
82074  [[50590067, 50592951]]    [[2884]]  [2884]   
82075  [[50590067, 50592951]]    [[2884]]  [2884]   
82076  [[50590067, 50592951]]    [[2884]]  [2884]   
82077  [[50590067, 50592951]]    [[2884]]  [2884]   
82078  [[50590067, 50592951]]    [[288

해당 colums의 요소를 확인하고 싶을 경우 불러내어 np.unique를 쓰면 중복을 제거하여 보여줍니다

In [32]:
np.unique(refFlat_DF['transName'])

array(['A1BG', 'A1BG-AS1', 'A1CF', ..., 'ZYX', 'ZZEF1', 'ZZZ3'],
      dtype=object)

처리한 테이블 데이터를 저장하려면 csv같은 형식도 가능하나, 이는 아무래도 자료형이나 분할까지 적용하기에는 어려우며 직접 보는 것도 힘드므로  
아에 바이너리 파일로 저장하는 것이 편하기도 합니다

In [33]:
refFlat_DF.to_pickle('./refFlat.pkl')

불러올 땐 다음과 같습니다

In [34]:
refFlat_DF_Re = pd.read_pickle('./refFlat.pkl')

In [35]:
refFlat_DF_Re

transName       transID                 chrom strand    txnSta  \
0        DDX11L1     NR_046018                  chr1      +     11873   
1         WASH7P     NR_024540                  chr1      -     14361   
2      MIR6859-1     NR_106918                  chr1      -     17368   
3      MIR6859-2     NR_107062                  chr1      -     17368   
4      MIR6859-3     NR_107063                  chr1      -     17368   
...          ...           ...                   ...    ...       ...   
82077      GRB10  NM_001001555                  chr7      -  50590067   
82078      GRB10  NM_001371008                  chr7      -  50590067   
82079      GRB10  NM_001350816                  chr7      -  50590068   
82080  RAB11FIP3  NM_001142272                 chr16      +    474923   
82081  RAB11FIP3  NM_001142272  chr16_KI270854v1_alt      +     75229   

         txnEnd    cdsSta    cdsEnd  exnNum  \
0         14409     14409     14409       3   
1         29370     29370     29370      11   
2         17436     17436     17436       1   
3         17436     17436     17436       1   
4         17436     17436     17436       1   
...         ...       ...       ...     ...   
82077  50793453  50592951  50674623      17   
82078  50793453  50592951  50674623      16   
82079  50792874  50592951  50674623      18   
82080    523011    482551    520839      12   
82081    125679     82856    123507      12   

                                                  exnSta  ...  \
0                                  [11873, 12612, 13220]  ...   
1      [14361, 14969, 15795, 16606, 16857, 17232, 176...  ...   
2                                                [17368]  ...   
3                                                [17368]  ...   
4                                                [17368]  ...   
...                                                  ...  ...   
82077  [50590067, 50595436, 50603997, 50604310, 50605...  ...   
82078  [50590067, 50595436, 50603997, 50604310, 50605...  ...   
82079  [50590068, 50595436, 50603997, 50604310, 50605...  ...   
82080  [474923, 482524, 488850, 496823, 503003, 50552...  ...   
82081  [75229, 82829, 89154, 99694, 105890, 108215, 1...  ...   

                                                    utr5  \
0                                                     []   
1                                                     []   
2                                                     []   
3                                                     []   
4                                                     []   
...                                                  ...   
82077  [[50674623, 50674658], [50703820, 50703908], [...   
82078  [[50674623, 50674658], [50732271, 50732368], [...   
82079  [[50674623, 50674658], [50703820, 50703908], [...   
82080               [[474923, 475045], [482524, 482551]]   
82081                   [[75229, 75351], [82829, 82856]]   

                             utr5LenList utr5Len  \
0                                     []       0   
1                                     []       0   
2                                     []       0   
3                                     []       0   
4                                     []       0   
...                                  ...     ...   
82077         [[35], [88], [170], [230]]   [523]   
82078                [[35], [97], [230]]   [362]   
82079  [[35], [88], [116], [170], [433]]   [842]   
82080                      [[122], [27]]   [149]   
82081                      [[122], [27]]   [149]   

                                                 cdsList  \
0       [[11873, 12227], [12612, 12721], [13220, 14409]]   
1      [[14361, 14829], [14969, 15038], [15795, 15947...   
2                                       [[17368, 17436]]   
3                                       [[17368, 17436]]   
4                                       [[17368, 17436]]   
...                                                  ...   
82077  [[50592951,